mdsa/algorithms/scmds.py

```diff
@@ -0,0 +1,402 @@
+        """Returns an affine mapping function.
+
+        Arguments are two different MDS solutions/mappings (identical
+        dimensions) on the the same objects/distances.
+
+        Affine mapping function:
+        Y = UX + kron(b,ones(1,n)), UU' = I
+        X = [x_1, x_2, ... , x_n]; x_j \in R^m
+        Y = [y_1, y_2, ... , y_n]; y_j \in R^m
+
+        From Tzeng 2008:
+        `The projection of xi,2 from q2 dimension to q1 dimension
+        induces computational errors (too). To avoid this error, the
+        sample number of the overlapping region is important. This
+        sample number must be large enough so that the derived
+        dimension of data is greater or equal to the real data`
+
+        Notes:
+        - This was called moving in scmdscale.m
+        - We work on tranposes, i.e. coordinates for objects are in columns
+
+
+        Arguments:
+        - `matrix_x`: first mds solution (`reference`)
+        - `matrix_y`: seconds mds solution
+
+        Return:
+        A tuple of unitary operator u and shifting operator b such that:
+        y = ux + kron(b, ones(1,n))
+        """
+
+        if matrix_x.shape != matrix_y.shape:
+            raise ValueError("input matrices are not of same size")
+
+        if not matrix_x.shape[0] <= matrix_x.shape[1]:
+            raise ValueError("input matrices should have more columns than rows")
+
+        # Have to check if we have not more rows than columns, otherwise,
+        # the qr function below might behave differently in the matlab
+        # prototype, but not here. In addition this would mean that we
+        # have more dimensions than overlapping objects which shouldn't
+        # happen
+        #
+        # matlab code uses economic qr mode (see below) which we can't use
+        # here because we need both return matrices.
+        #
+        # see
+        # http://www.mathworks.com/access/helpdesk/help/techdoc/index.html?/access/helpdesk/help/techdoc/ref/qr.html
+        # [Q,R] = qr(A,0) produces the economy-size decomposition.
+        # If m > n, only the first n columns of Q and
+        # the first n rows of R are computed.
+        # If m<=n, this is the same as [Q,R] = qr(A)
+        #
+        # [Q,R] = qr(A), where A is m-by-n, produces
+        # an m-by-n upper triangular matrix R and
+        # an m-by-m unitary matrix Q so that A = Q*R
+        #
+        # That's why we check above with an assert
+
+        ox = self.rowmeans(matrix_x)
+        oy = self.rowmeans(matrix_y)
+
+        mx = matrix_x - kron(ox, ones((1, matrix_x.shape[1])))
+        my = matrix_y - kron(oy, ones((1, matrix_x.shape[1])))
+
+        (qx, rx) = qr(mx)
+        (qy, ry) = qr(my)
+
+        # sign correction
+        #
+        # Daniel suggest to use something like
+        # [arg]where(sign(a.diagonal()) != sign(b.diagonal())) and then
+        # iterate over the results. Couldn't figure out how to do this
+        # properly :(
```

Comment:
zeros playing hell? can always cython later if this method is a performance bottleneck, in which case a large amount of benefit is gained by unrolling loops as much as possible